Add new track pieces to the Twister, Hyper Twister and Vertical Drop Roller Coasters

mixiate opened the pull request:
This adds new track pieces to the twister, hyper twister and vertical drop rollercoasters.

 - Large unbanked and banked sloped curves (Twister, Hyper, Vertical Drop)
 - Large corkscrews (Twister)
 - Medium half loops (Twister, Vertical Drop)
 - Small and large zero g rolls (Cheats)

![newtwisterpieces](https://github.com/user-attachments/assets/e9b85a44-bd25-4763-9f35-ff905c661551)

All pieces have been tested with the applicable trains. The zero g rolls are currently cheats only as they don't seem to currently be optionally available only for trains that have the sprites. The trains don't have sprites for all the diagonal banking pieces either but these are already available for the giga coaster so I just made them available to these too. Maybe both of these things could be sorted out eventually.

SpartanFrederic104 commented:
Did some testing with these new track pieces and they just look fantastic and smooth as usual.
![Fred's Workbench 2024-10-25 13-47-09](https://github.com/user-attachments/assets/bf901cb9-c7fb-401c-bb50-4d8b145f6e12)
![Fred's Workbench 2024-10-25 13-47-15](https://github.com/user-attachments/assets/e6ccd527-610f-4582-9cd3-aa65ae052d47)
![Fred's Workbench 2024-10-25 13-47-31](https://github.com/user-attachments/assets/51532f3e-7b30-4eb5-9ebb-c5e6bc2635bc)
I haven't found any major issues so far (well, excluding the vanilla trains glitching on some diagonal banked slope pieces and the lack of sprites for zero-G rolls). Can't wait to play with these new track pieces in a future release build!
(and yes these screenshots were taken before the force push was done)

mixiate commented:
Thanks for testing them again, I really appreciate it. I just fixed a few pieces that didn't connect quite right, and a couple of supports that didn't reach, and rebased it. Hopefully this is all okay now.

fidwell commented:
These look great. The corkscrews are even better looking than the originals. Nice work!

![Dorney Park   Wildwater Kingdom 2024-10-25 18-46-27](https://github.com/user-attachments/assets/f3126ad1-671a-4eea-8ec7-6b31f99f17e3)

karst commented:
Thanks for taking in my feedback regarding the polygon spines :)

karst commented:
There's a few more things required before this can be merged:
- Network version needs to be increased
- There needs to be conversion code for making the new track pieces invisible on rides that already have them before this was added to the game. As an example check this: https://github.com/OpenRCT2/OpenRCT2/pull/17997/files#diff-5589e6ccdd14fc6e0ff9416f0dbd6b7bff23033380e8649da2336b84f0382e20R921-R937
- To make this happen you also need to bump the park file version.

mixiate commented:
Yep, i'll sort this out as soon as I can.

Gymnasiast commented on src/openrct2/ride/rtd/coaster/VerticalDropCoaster.h:
```diff
@@ -23,8 +23,8 @@ constexpr RideTypeDescriptor VerticalDropCoasterRTD =
-        .enabledTrackGroups = {TrackGroup::flat, TrackGroup::straight, TrackGroup::stationEnd, TrackGroup::liftHill, TrackGroup::liftHillSteep, TrackGroup::flatRollBanking, TrackGroup::verticalLoop, TrackGroup::slope, TrackGroup::slopeSteepUp, TrackGroup::slopeSteepDown, TrackGroup::flatToSteepSlope, TrackGroup::slopeCurve, TrackGroup::slopeCurveSteep, TrackGroup::sBend, TrackGroup::curveSmall, T ...
-        .extraTrackGroups = {TrackGroup::halfLoop, TrackGroup::corkscrew, TrackGroup::barrelRoll, TrackGroup::poweredLift, TrackGroup::halfLoopLarge, TrackGroup::quarterLoop, TrackGroup::booster, TrackGroup::twist},
+        .enabledTrackGroups = {TrackGroup::flat, TrackGroup::straight, TrackGroup::stationEnd, TrackGroup::liftHill, TrackGroup::liftHillSteep, TrackGroup::flatRollBanking, TrackGroup::verticalLoop, TrackGroup::slope, TrackGroup::slopeSteepUp, TrackGroup::slopeSteepDown, TrackGroup::flatToSteepSlope, TrackGroup::slopeCurve, TrackGroup::slopeCurveSteep, TrackGroup::sBend, TrackGroup::curveSmall, T ...
+        .extraTrackGroups = {TrackGroup::halfLoop, TrackGroup::corkscrew, TrackGroup::barrelRoll, TrackGroup::poweredLift, TrackGroup::halfLoopLarge, TrackGroup::quarterLoop, TrackGroup::booster, TrackGroup::twist, TrackGroup::zeroGRoll, TrackGroup::zeroGRollLarge, TrackGroup::corkscrewLarge},
```
Is there a particular reason to lock the Zero G Rolls behind a cheat?

karst replied:
imo there isn't, if a vehicle can't draw the pieces they'll be disabled anyhow, and the vertical drop coaster is known to have this element, especially immelmen.

mixiate replied:
Originally it was because none of the new inversions had entries for the sprites required, but that has now been fixed. I added the zero g rolls to the twister but forgot about updating these for the vertical drop coaster. Should the large corkscrew also be available?

karst replied:
Yeah definitely

Realsteel89 replied:
yes and i made a lot of BM Cars now but i think we need an permission to make them official?

Gymnasiast replied:
@mixiate Ok, could you update and rebase?

karst commented:
As far as I can see the code looks all good. I'll have a dev look at it tho.
I'll give this a playtest.

karst commented on src/openrct2/park/Legacy.cpp:
```diff
@@ -2563,6 +2563,82 @@ bool TrackTypeMustBeMadeInvisible(ride_type_t rideType, OpenRCT2::TrackElemType
+            case TrackElemType::LeftLargeHalfLoopUp:
+            case TrackElemType::RightLargeHalfLoopUp:
+            case TrackElemType::LeftLargeHalfLoopDown:
+            case TrackElemType::RightLargeHalfLoopDown:
```
These four lines need to be removed. Sending a patch right away.
